### PR TITLE
Let argument parser handle enum conversions

### PR DESCRIPTION
**Pull request by djc**

### CODE REVIEW

File: src/dist/dist.rs [hunk 1]
```diff
@@ -611,6 +613,42 @@ pub enum Profile {
     Complete,
 }
 
+impl Profile {
+    pub(crate) fn names(sep: &str) -> String {
```
Comment (rami3l) [inline src/dist/dist.rs]: You can also use https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.join for this like in https://github.com/rust-lang/rustup/pull/3874#discussion_r1641180721. 

Also, maybe it's a good idea to unify the names of both methods for the sake of coherence (e.g. `Self::join`), or, maybe we can just inline it if it's just a single-use method that can be reduced to a single line?

File: src/cli/self_update.rs [hunk 2]
```diff
@@ -100,21 +102,48 @@ pub(crate) const NEVER_SELF_UPDATE: bool = true;
...
+    pub(crate) fn join(sep: &str) -> String {
+        let mut buf = String::new();
+        for mode in SelfUpdateMode::value_variants() {
+            if !buf.is_empty() {
+                buf.push_str(sep);
+            }
+            buf.push_str(mode.as_str());
+        }
+
+        buf
```
Comment (rami3l) [inline src/cli/self_update.rs]: You can use https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.join for this if you'd prefer, seems easier to parse to me :)

```suggestion
        Self::value_variants().iter().join(sep)
```

File: src/config.rs [hunk 3]
```diff
@@ -425,11 +415,10 @@ impl Cfg {
 
     pub(crate) fn get_self_update_mode(&self) -> Result<SelfUpdateMode> {
         self.settings_file.with(|s| {
-            let mode = match &s.auto_self_update {
-                Some(mode) => mode.clone(),
+            Ok(match s.auto_self_update {
```
Comment (rami3l) [inline src/config.rs:413]: Nit: How about using one of the `.unwrap_or*` methods here?

**Review by rami3l — APPROVED** [review]
This is exactly the kind of improvement I'd like to see after migrating to `clap_derive`, many thanks!

File: src/dist/dist.rs (hunk 1 above)
Comment (djc) [reply src/dist/dist.rs]: Fair, changed it!

File: src/cli/self_update.rs (hunk 2 above)
Comment (djc) [reply src/cli/self_update.rs]: Good call, changed it.

File: src/config.rs (hunk 3 above)
Comment (djc) [reply src/config.rs:413]: Yeah, much better!

**Pull request merged**